nucleus/deploy/nucleus_integration.py

```diff
@@ -0,0 +1,156 @@
+            ), "If you don't pass a nucleus model you better pass a model name and reference id"
+            model = self.nucleus_client.add_model(
+                name=model_name, reference_id=model_ref_id
+            )
+
+        # Create a Nucleus model run
+        model_run = model.create_run(
+            name=model_run_name, dataset=self.dataset, predictions=[]
+        )
+        prediction_items = []
+        for s3url, dataset_item in self.s3url_to_dataset_map.items():
+            item_link = self.hmi_async_job.responses[s3url]
+            print(f"item_link={item_link}")
+            # e.g. s3://scale-ml/tmp/hosted-model-inference-outputs/a224499e-50ac-4b08-ad0c-c18e74c14184.pkl
+            kwargs = {
+                "transport_params": {
+                    "session": Session(profile_name="ml-worker")
+                }
+            }
+
+            with smart_open.open(item_link, "rb", **kwargs) as bundle_pkl:
+                inference_result = cloudpickle.load(bundle_pkl)
+                ref_id = dataset_item.reference_id
+                for box in inference_result:
+                    # TODO assuming box is a list of (x, y, w, h, label). This is almost certainly not the case.
+                    # We will have to use a ModelEndpoint/ModelBundle that returns boxes in this format.
+                    # Also, label is probably returned as an integer instead of a label that makes semantic sense
+                    pred_item = nucleus.BoxPrediction(
+                        label=box["label"],
+                        x=box["left"],
+                        y=box["top"],
+                        width=box["width"],
+                        height=box["height"],
+                        reference_id=ref_id,
+                    )
+                    prediction_items.append(pred_item)
+
+        job = model_run.predict(prediction_items, asynchronous=True)
+        job.sleep_until_complete()
+        job.errors()
+
+
+def create_nucleus_dataset_inference_run(
+    hmi_endpoint: ModelEndpoint, nucleus_client, dataset: Dataset
+):
+    """
+    Returns a NucleusDatasetInferenceRun, client will need to periodically call poll on this in order to upload
+    """
+    s3urls, s3url_to_dataset_map = _nucleus_ds_to_s3url_list(dataset)
+    async_job = hmi_endpoint.predict(s3urls)
+    return NucleusDatasetInferenceRun(
+        async_job, nucleus_client, s3url_to_dataset_map, dataset
+    )
+
+
+def _nucleus_ds_to_s3url_list(
+    dataset: Dataset,
+) -> Tuple[List[str], Dict[str, DatasetItem]]:
+    # TODO I'm not sure if dataset items are necessarily s3URLs. Does this matter?
+    # TODO support lidar point clouds
+    if len(dataset.items) == 0:
+        logger.warning("Passed a dataset of length 0")
+        return [], {}  # TODO return type?
+    dataset_item_type = dataset.items[0].type
+    if not all(data.type == dataset_item_type for data in dataset.items):
+        logger.warning("Dataset has multiple item types")
```

Comment:
This is going to be normal for scene datasets that have a mix of pointcloud and image data. Our approach on the autotag/model zoo side has been to use only the image data. But you may want to think of this more carefully in case people want to host models that are 3D.

Reply:
Good point, we'll need to flesh out how Nucleus interacts with HMI a bit more I think. Possibly one thing we could do is have users define if an HMI ModelBundle/ModelEndpoint does 2D, 3D, or general inference, where if a ModelBundle is 2D or 3D, then it needs to conform to a pretty strict API. Otherwise, we'd somehow have to automatically translate between data types (e.g. detect when a bbox is (x, y, w, h) or (x1, y1, x2, y2)) which sounds close to impossible.

Reply:
I think both the bundle and endpoint would be a black box, i.e. it's up to the user to define if it does 2D, 3D, or mixed inference, without needing explicit structured tagging from them. They'd also be responsible for providing the preprocessing/postprocessing code to translate into the necessary Nucleus python client types, which should be doable since the requests in this context are originating from the Nucleus client.